src/main/java/at/hannibal2/skyhanni/utils/EntityOutlineRenderer.kt

```diff
@@ -0,0 +1,398 @@
+            val mc = Minecraft.getMinecraft()
+
+            // Only render the view entity when sleeping or in 3rd person mode mode
+            return if (entity === mc.renderViewEntity &&
+                    !(mc.renderViewEntity is EntityLivingBase && (mc.renderViewEntity as EntityLivingBase).isPlayerSleeping ||
+                            mc.gameSettings.thirdPersonView != 0)) {
+                false
+            } else mc.theWorld.isBlockLoaded(BlockPos(entity)) && (mc.renderManager.shouldRender(entity, camera, x, y, z) || entity.riddenByEntity === mc.thePlayer)
+            // Only render if renderManager would render and the world is loaded at the entity
+        }
+
+        private val BUF_FLOAT_4: java.nio.FloatBuffer = org.lwjgl.BufferUtils.createFloatBuffer(4)
+
+        private fun outlineColor(color: Int) {
+            BUF_FLOAT_4.put(0, (color shr 16 and 255).toFloat() / 255.0f)
+            BUF_FLOAT_4.put(1, (color shr 8 and 255).toFloat() / 255.0f)
+            BUF_FLOAT_4.put(2, (color and 255).toFloat() / 255.0f)
+            BUF_FLOAT_4.put(3, 1f)
+            GL11.glTexEnv(GL11.GL_TEXTURE_ENV, GL11.GL_TEXTURE_ENV_COLOR, BUF_FLOAT_4)
+        }
+
+        /**
+         * Function that copies a portion of a framebuffer to another framebuffer.
+         *
+         *
+         * Note that this requires GL3.0 to function properly
+         *
+         *
+         * The major use of this function is to copy the depth-buffer portion of the world framebuffer to the entity outline framebuffer.
+         * This enables us to perform no-xray outlining on entities, as we can use the world framebuffer's depth testing on the outline frame buffer
+         *
+         * @param frameToCopy   the framebuffer from which we are copying data
+         * @param frameToPaste  the framebuffer onto which we are copying the data
+         * @param buffersToCopy the bit mask indicating the sections to copy (see [GL11.GL_DEPTH_BUFFER_BIT], [GL11.GL_COLOR_BUFFER_BIT], [GL11.GL_STENCIL_BUFFER_BIT])
+         */
+        private fun copyBuffers(frameToCopy: Framebuffer?, frameToPaste: Framebuffer?, buffersToCopy: Int) {
+            if (OpenGlHelper.isFramebufferEnabled()) {
+                OpenGlHelper.glBindFramebuffer(GL30.GL_READ_FRAMEBUFFER, frameToCopy!!.framebufferObject)
+                OpenGlHelper.glBindFramebuffer(GL30.GL_DRAW_FRAMEBUFFER, frameToPaste!!.framebufferObject)
+                GL30.glBlitFramebuffer(0, 0, frameToCopy.framebufferWidth, frameToCopy.framebufferHeight,
+                        0, 0, frameToPaste.framebufferWidth, frameToPaste.framebufferHeight,
+                        buffersToCopy, GL11.GL_NEAREST)
+            }
+        }
+
+        fun isCacheEmpty(): Boolean {
+            return isXrayCacheEmpty() && isNoXrayCacheEmpty()
+        }
+
+        private fun isXrayCacheEmpty(): Boolean {
+            return entityRenderCache.xrayCache == null || entityRenderCache.xrayCache!!.isEmpty()

```

Comment:
Kotlin has `.?` and `?:` that could be used here: `entityRenderCache.xrayCache?.isEmpty() ?: true`. See https://kotlinlang.org/docs/null-safety.html#elvis-operator for more infos